cli/server.go

```diff
@@ -127,6 +127,19 @@ func server() *cobra.Command {
 			ctx, cancel := context.WithCancel(cmd.Context())
 			defer cancel()
 
+			// Register signals early on so that graceful shutdown can't
+			// be interrupted by additional signals. Note that we avoid
+			// shadowing cancel() (from above) here because notifyStop()
+			// restores default behavior for the signals. This protects
+			// the shutdown sequence from abrubtly terminating things
+			// like: database migrations, provisioner work, workspace
+			// cleanup in dev-mode, etc.
+			//
+			// To get out of a graceful shutdown, the user can send
+			// SIGQUIT with ctrl+\ or SIGKILL with `kill -9`.
+			notifyCtx, notifyStop := signal.NotifyContext(ctx, interruptSignals...)
```

Comment:
Note: Naming this `notifyCtx` here means that during startup, and interrupt signal won't abort the process up until we start listening `<-notifyCtx.Done()`.

The reason for this is that we have certain services that begin their shutdown procedure on context cancellation, but we want to control the order in which things are shut down.

I think it's acceptable that after starting, ctrl+c won't close `coder server` until the initial startup sequence is complete. But it's possible to work around this by monitoring the `notifyCtx` separately for parts of the startup where we want an early exit.